activesupport/lib/active_support/backtrace_cleaner.rb

```diff
@@ -104,27 +104,33 @@ def add_stdlib_silencer
         add_silencer { |line| line.start_with?(RbConfig::CONFIG["rubylibdir"]) }
       end
 
-      def filter_backtrace(backtrace)
-        @filters.each do |f|
-          backtrace = backtrace.map { |line| f.call(line) }
+      # Returns the original collection if the block returns an empty collection.
+      def empty_guard(collection)
+        result = yield
+        result.empty? ? collection.dup : result
+      end
+
+      # process +ary+ via +filters+ using +method+, ensuring
+      # _something_ gets returned.
+      def process_collection(ary, filters, method)
+        empty_guard ary do
+          filters.reduce(ary) { |bt, f| bt.send(method) { |line| f.call(line) } }
         end
+      end
 
-        backtrace
+      # use @filters to transform the backtrace via map
+      def filter_backtrace(backtrace)
+        process_collection backtrace, @filters, :map
       end
 
+      # use @silencers to reject parts of the backtrace
       def silence(backtrace)
-        @silencers.each do |s|
-          backtrace = backtrace.reject { |line| s.call(line) }
-        end
-
-        backtrace
+        process_collection backtrace, @silencers, :reject
```

Comment:
If this is the only spot where `empty_guard` is necessary, perhaps it could be rolled into this method:

```ruby
keep = @silencers.reduce(backtrace) do |bt, s|
  bt.reject { |line| s.call(line) }
end

keep.first ? keep : backtrace
```

The above code also sidesteps the issue of Enumerator not responding to `empty?`.

Reply:
I like the `.first` change... 

I gotta ask, what's the point of making a known collection lazy? You already have the collection and things like `find` and `any?` are already shortcutting. That's why I took the `lazy` out. Totally fine with putting it back.

Reply:
In this case, using `first`, it avoids processing the entire collection.  It looks like the `lazy` was added in #35985 as a performance optimization.  Since it was for performance reasons, it would be nice to avoid calling `first` twice, but I don't see a way around that.